app/services/idv/socure_user_set.rb

```diff
@@ -4,15 +4,34 @@ module Idv
   class SocureUserSet
     attr_reader :redis_pool
 
+    ADD_USER_SCRIPT = <<~LUA_EOF
+      local key = KEYS[1]
+      local user_uuid = ARGV[1]
+      local max_allowed_users = tonumber(ARGV[2])
+
+      local number_of_socure_users = redis.call('SCARD', key)
+      if number_of_socure_users >= max_allowed_users then
+        return false
+      end
+      redis.call('SADD', key, user_uuid)
+      return true
+    LUA_EOF
+
+    ADD_USER_SCRIPT_SHA1 = Digest::SHA1.hexdigest(ADD_USER_SCRIPT).freeze
+
     def initialize(redis_pool: REDIS_POOL)
       @redis_pool = redis_pool
     end
 
     def add_user!(user_uuid:)
-      return if maxed_users?
-
+      script_args = [user_uuid.to_s, IdentityConfig.store.doc_auth_socure_max_allowed_users.to_i]
       redis_pool.with do |client|
-        client.sadd(key, user_uuid)
+        begin
+          client.evalsha(ADD_USER_SCRIPT_SHA1, [key], script_args)
+        rescue Redis::CommandError => error
+          raise error unless error.message.start_with?('NOSCRIPT')
+          client.eval(ADD_USER_SCRIPT, [key], script_args)
+        end
```

Comment:
WDYT of loading this script into Redis via a Rails initializer to make performance a little more deterministic

something like...


```diff
+++ b/config/initializers/01_redis.rb
@@ -6,6 +6,9 @@ REDIS_POOL = ConnectionPool.new(size: IdentityConfig.store.redis_pool_size) do
   Redis.new(url: IdentityConfig.store.redis_url)
 end.freeze
 
+require 'idv/socure_user_set'
+Idv::SocureSet.cache_user_script!(redis_pool: REDIS_POOL)
+
```

and then maybe like


```ruby
def self.cache_user_script!(redis_pool: REDIS_POOL)
  redis_pool.with do |redis|
    next if redis.script(:exists, ADD_USER_SCRIPT_SHA1)
    redis.script(:load, ADD_USER_SCRIPT)
  end
end
```

So all this could simplify to
```suggestion
        client.evalsha(ADD_USER_SCRIPT_SHA1, [key], script_args)
```

Reply:
I think there's no guarantee that `evalsha` succeeds, even if it is cached ahead of time: https://redis.io/docs/latest/develop/interact/programmability/eval-intro/#cache-volatility

Reply:
Ah dang what a bummer!